projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    const levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+    const nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+    const nodeCoOrdinatesMap: Map<string, TopologyCoordinates> = new Map<string, TopologyCoordinates>();
+
+    this.findRootNodes(topology, levelToNodesMap, nodeToLevelMap);
+    this.fillNodeAndLevelMaps(levelToNodesMap, nodeToLevelMap);
+    this.fillNodeCoOrdinatesMap(nodeCoOrdinatesMap, levelToNodesMap);
+    this.setNodeCoOrdinates(topology, nodeCoOrdinatesMap);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>, levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        nodeToLevelMap.set(node, 0);
+        levelToNodesMap.has(0) ? levelToNodesMap.get(0)?.push(node) : levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps(levelToNodesMap: Map<number, RenderableTopologyNode[]>, nodeToLevelMap: Map<RenderableTopologyNode, number>) {
+
+    const rootNodes: RenderableTopologyNode[] = levelToNodesMap.get(0) ?? [];
+    const visited = new Set<string>();
+    let curr = 0;
+
+    while (curr >= 0) {
```

Comment:
can we not do `curr < rootNodes.length`, instead of breaking it later?

Reply:
won't it calculate the length once, i.e., in the starting and use the same length for the entire loop?

Reply:
Ahh, you're updating the rootNodes in the loop

Reply:
updated, it do check the length each time